File: src/main/java/proto/games/matchmaker/domain/matchmaker/position/PositionMatchmaker.java
package proto.games.matchmaker.domain.matchmaker.position;

import proto.games.matchmaker.domain.matchmaker.PlayerBulk;
import proto.games.matchmaker.domain.matchmaker.rank.RankTeam;
import proto.games.matchmaker.domain.player.Player;

import java.util.List;
import java.util.stream.Collectors;

public class PositionMatchmaker {

    public List<PositionTeam> match(PlayerBulk playerBulk, List<RankTeam> rankTeams) {
        List<PositionTeam> positionTeams = prepare(rankTeams);
        distribute(positionTeams, playerBulk);
        populateMissingPositions(positionTeams, playerBulk);
        return positionTeams;
    }

    private List<PositionTeam> prepare(List<RankTeam> rankTeams) {
        return rankTeams.stream()
                .map(PositionTeam::new)
                .collect(Collectors.toList());
    }

    private void distribute(List<PositionTeam> teams, PlayerBulk playerBulk) {
        for (PositionTeam team : teams) {
            for (Player player : playerBulk.getPlayers()) {
                if (team.include(player)) {
                    playerBulk.remove(player);
                }
            }
        }
    }

    private void populateMissingPositions(List<PositionTeam> teams, PlayerBulk playerBulk) {
        for (PositionTeam team : teams) {
            for (Player player : playerBulk.getPlayers()) {
                if (team.populate(player)) {
                    playerBulk.remove(player);
                }
            }
        }
    }
}

File: src/main/kotlin/proto/games/matchmaker/domain/matchmaker/position/PositionMatchmaker.kt
package proto.games.matchmaker.domain.matchmaker.position

import proto.games.matchmaker.domain.matchmaker.PlayerBulk
import proto.games.matchmaker.domain.matchmaker.rank.RankTeam

class PositionMatchmaker {
    fun match(playerBulk: PlayerBulk, rankTeams: List<RankTeam>): List<PositionTeam> {
        val positionTeams = prepare(rankTeams)
        distribute(positionTeams, playerBulk)
        populateMissingPositions(positionTeams, playerBulk)
        return positionTeams
    }

    private fun prepare(rankTeams: List<RankTeam>): List<PositionTeam> {
        return rankTeams
            .map { rankTeam -> PositionTeam(rankTeam) }
    }

    private fun distribute(teams: List<PositionTeam>, playerBulk: PlayerBulk) {
        for (team in teams) {
            for (player in playerBulk.players) {
                if (team.include(player)) {
                    playerBulk.remove(player)
                }
            }
        }
    }

    private fun populateMissingPositions(teams: List<PositionTeam>, playerBulk: PlayerBulk) {
        for (team in teams) {
            for (player in playerBulk.players) {
                if (team.populate(player)) {
                    playerBulk.remove(player)
                }
            }
        }
    }
}